### ISSUE TITLE
Using RoBERTa or LongFormer for texts with 16K tokens

### ISSUE THREAD

**Issue by vr25** [allenai/longformer#62]
> LongFormer does it by pooling all the local attentions (512) together in global attention (512 x 8 = 4096).

This is not entirely true. There's no "pooling" of the 4096 tokens into 512. We keep all 4096 tokens. The only change is how attention is computed; instead of every token attending to every other token, we change it such that every token attends to a smaller number of surrounding tokens. This speeds up selfattention computation (which is the bottleneck) by assuming that the attention score between certain pairs of words is zero. This doesn't change the architecture or introduce any pooling. 

We are working on some code that will make it easy to train your own long model, so you can try longer sequences. We know it is easy to get to 16K or even 32k with RoBERTa-base architecture (need base model, fp16, gradient checkpointing). For sequences longer than that, you will need to find ways to save memory depending on your application. For example, reducing window size, reducing size of the feed forward layers, implementing reversible transformers, use sinusoidal position embedding instead of learned position embedding.

_Originally posted by @ibeltagy in https://github.com/allenai/longformer/issues/48#issuecomment-634270401_

**Comment by vr25** [comment]
Hi @ibeltagy ,

Just a follow-up question.

I am now trying to use RoBERTa (and later LongFormer) on the texts with ~16K - 20K length (much shorter than 100K as of now). I want to know how can I use either of the two models for text classification and if I need to break down the entire document into chunks?

Any link to an example would be helpful.

Thanks!

**Comment by ibeltagy** [comment]
For text classification in general, check this answer: https://github.com/allenai/longformer/issues/34#issuecomment-638711390

One option to deal with sequences longer than the limit is to develop a new version of the model that supports longer sequences. This notebook shows how to do so https://github.com/allenai/longformer/blob/master/scripts/convert_model_to_long.ipynb.

If you want to classify the text without changing the model, you can try one of the following: 

- Truncate the document and only use the first 512 (or 4096) tokens

- Chuck the document into blocks of size 512 (or 4096), run each chunk through your model, get the embedding of the `[CLS]` token of each chunk, then pool these embeddings. This gives you a fixed-length vector for the input document that you can use for classification. This is how our text-classification RoBERTa baseline works.

**Comment by vr25** [comment]
And pooling would be taking a mean of all chunk vectors?

Also, thank you for the detailed explanation.

**Comment by ibeltagy** [comment]
mean or maxpool, I would expect both to work equally well.

**Comment by vr25** [comment]
I see, thanks! And use the same chunking and pooling strategy for fine-tuning (task) as well?

**Comment by ibeltagy** [comment]
>  for fine-tuning (task) as well?

I thought this is for fine-tuning (training on a classification dataset). What other forms of training are you also doing?

**Comment by vr25** [comment]
I meant to ask domain fine-tune (no labeled data), more like pre-training but not from scratch, instead just using the existing model to fit on the domain corpus, and/or then maybe task fine-tune (if labeled data is present).

But I think you've already answered my question, both would use the same strategy.

**Comment by vr25** [comment]
> For text classification in general, check this answer: [#34 (comment)](https://github.com/allenai/longformer/issues/34#issuecomment-638711390)
> 
> One option to deal with sequences longer than the limit is to develop a new version of the model that supports longer sequences. This notebook shows how to do so https://github.com/allenai/longformer/blob/master/scripts/convert_model_to_long.ipynb.
> 
> If you want to classify the text without changing the model, you can try one of the following:
> 
> * Truncate the document and only use the first 512 (or 4096) tokens
> * Chuck the document into blocks of size 512 (or 4096), run each chunk through your model, select the embedding of the `[CLS]` token from each chunk, then pool these embeddings. This gives you a fixed-length vector for the input document that you can use for classification. This is how our text-classification RoBERTa baseline works.

Hi @ibeltagy ,

1. I was trying to find code for the RoBERTa-base text classification where you've implemented the chunking and pooling strategy.

2. All I could find were the slice and stride techniques in [triviaqa.py](https://github.com/allenai/longformer/blob/master/scripts/triviaqa.py) and I am not sure if 1. and 2. are related conceptually.

Also, following the famous [post](https://mccormickml.com/2019/07/22/BERT-fine-tuning/) by Chris McCormick on classification fine-tuning, I am trying to figure out how I can break down the document at the encoding step.

In simple terms, say, I have a document of ```length 1324```, I want to use BERT/RoBERTa. If I want to chunk it, then I will have ```512, 512 and 300 (padded to 512)``` length chunks each with a ```[CLS]``` token at the beginning. The next step is to pool all these ```[CLS]``` tokens to get one final ```[CLS]``` for that document.

Now that I am actually implementing it, I am a little unsure if the pooling should be done
i. before/after the backward pass
ii. after the training and before validation.

Thank you, again!

**Comment by ibeltagy** [comment]
>   and I am not sure if 1. and 2. are related conceptually.

Conceptually, they are (both employ some type of chunking), but the details are different. The classification one should be much simpler than QA.

> i. before/after the backward pass

before the backward pass. The pooling is part of how you are building the document representation. 

>  after the training and before validation.

pooling happens in both, training and valuation. It is part of the forward function.

**Comment by ibeltagy** [comment]
Closing. Please feel free to reopen or create a new one if you have more questions.

**Comment by vr25** [comment]
@ibeltagy 

Did you fine-tune RoBERTa (unfreeze all 12 layers) for creating document vectors using chunking-pooling logic?

**Comment by Saraharas** [comment]
Hi @ibeltagy ,

Following the conversation about chunking and pooling strategy for sentence classification, could you please give a hint how to use RoBERTa without changing the model for HotpotQA question answering with sequences longer than 512 tokens? Should the span prediction target be defined for each chunk separately?

Thanks!

**Comment by Saraharas** [comment]
Hi @ibeltagy ,

Could you please give a hint how to use RoBERTa without changing the model for HotpotQA question answering with sequences longer than 512 tokens? Should the span prediction target be defined for each chunk separately?

Thanks!
